File: 17.电话号码的字母组合.py
#
# @lc app=leetcode.cn id=17 lang=python3
#
# [17] 电话号码的字母组合
#

# @lc code=start
from typing import List


class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if digits == "":
            return []
        digits_dict = {
            '2': 'abc',
            '3': 'def',
            '4': 'ghi',
            '5': 'jkl',
            '6': 'mno',
            '7': 'pqrs',
            '8': 'tuv',
            '9': 'wxyz',
        }
        result = []

        def helper(i, s):
            # 终止条件
            if i == len(digits):
                result.append(s)
                return
            letters = digits_dict[digits[i]]
            for j in range(len(letters)):
                helper(i+1,  s+letters[j])
        helper(0, "")
        return result


s = Solution()
r = s.letterCombinations("236")
print('r ->', r)
# @lc code=end

File: 17.电话号码的字母组合.js
/*
 * @lc app=leetcode.cn id=17 lang=javascript
 *
 * [17] 电话号码的字母组合
 */

// @lc code=start
/**
 * @param {string} digits
 * @return {string[]}
 */
var letterCombinations = function (digits) {
  if (digits == "") {
    return []
  }
  let digitsDict = {
    '2': 'abc',
    '3': 'def',
    '4': 'ghi',
    '5': 'jkl',
    '6': 'mno',
    '7': 'pqrs',
    '8': 'tuv',
    '9': 'wxyz',
  }
  let result = []
  /**
   * @param {number} i 
   * @param {string} s 
   */
  function helper(i, s) {
    if (i === digits.length) {
      result.push(s)
      return
    }
    let chars = digitsDict[digits[i]]
    for (const char of chars) {
      helper(i + 1, s + char)
    }
  }
  helper(0, "")
  return result

};
// @lc code=end

